File: legacy/_1.12.2/src/main/java/cn/enaium/noexpensive/callback/AnvilSetOutputCallback.java
package cn.enaium.noexpensive.callback;

import net.legacyfabric.fabric.api.event.Event;
import net.legacyfabric.fabric.api.event.EventFactory;
import net.minecraft.item.ItemStack;

/**
 * @author Enaium
 */
public interface AnvilSetOutputCallback {
    Event<AnvilSetOutputCallback> EVENT = EventFactory.createArrayBacked(AnvilSetOutputCallback.class, (listeners) -> (output, levelCost, canTake) -> {
        for (AnvilSetOutputCallback listener : listeners) {
            listener.interact(output, levelCost, canTake);
        }
    });

    void interact(ItemStack output, int levelCost, boolean canTake);
}

File: legacy/_1.9.4/src/main/kotlin/cn/enaium/noexpensive/callback/AnvilSetOutputCallback.kt
package cn.enaium.noexpensive.callback

import net.legacyfabric.fabric.api.event.EventFactory
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack

/**
 * @author Enaium
 */
fun interface AnvilSetOutputCallback {
    fun interact(output: ItemStack?, levelCost: Int, canTake: Boolean, player: PlayerEntity)

    companion object {
        val EVENT = EventFactory.createArrayBacked(
            AnvilSetOutputCallback::class.java
        ) { listeners: Array<AnvilSetOutputCallback> ->
            AnvilSetOutputCallback { output: ItemStack?, levelCost: Int, canTake: Boolean, player: PlayerEntity ->
                for (listener in listeners) {
                    listener.interact(output, levelCost, canTake, player)
                }
            }
        }
    }
}
